cmd/aws-s3-csi-controller/csicontroller/reconciler.go

```diff
@@ -0,0 +1,284 @@
+		log.Error(err, "Failed to spawn Mountpoint Pod")
+	}
+	return nil
+}
+
+// spawnMountpointPod spawns a new Mountpoint Pod for given `pod` and volume.
+// The Mountpoint Pod will be spawned into the same node as `pod`, which then the mount operation
+// will be continued by the CSI Driver Node component in that node.
+func (r *Reconciler) spawnMountpointPod(
+	ctx context.Context,
+	pod *corev1.Pod,
+	pvc *corev1.PersistentVolumeClaim,
+	pv *corev1.PersistentVolume,
+	_ *corev1.CSIPersistentVolumeSource,
+	name string,
+) error {
+	log := logf.FromContext(ctx).WithValues(
+		"pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name},
+		"mountpointPod", name,
+		"pvc", pvc.Name, "volumeName", pv.Name)
+
+	log.Info("Spawning Mountpoint Pod")
+
+	mpPod := r.mountpointPodCreator.Create(pod, pvc)
+	if mpPod.Name != name {
+		err := fmt.Errorf("Mountpoint Pod name mismatch %s vs %s", mpPod.Name, name)
+		log.Error(err, "Name mismatch on Mountpoint Pod")
+		return err
+	}
+
+	err := r.Create(ctx, mpPod)
+	if err != nil {
+		log.Error(err, "Failed to create Mountpoint Pod")
+		return err
+	}
+
+	log.Info("Mountpoint Pod spawned", "mountpointPodUID", mpPod.UID)
+	return nil
+}
+
+// errPVCIsNotBoundToAPV is returned when given PVC is not bound to a PV yet.
+// This is not a terminal error - as PVCs can be bound to PVs dynamically - and just a transient error
+// to be retried later.
+var errPVCIsNotBoundToAPV = errors.New("PVC is not bound to a PV yet")
+
+// getBoundPVForPodClaim tries to find bound PV and PVC from given `claim`.
+// It `errPVCIsNotBoundToAPV` if PVC is not bound to a PV yet to be eventually retried.
+func (r *Reconciler) getBoundPVForPodClaim(
+	ctx context.Context,
+	pod *corev1.Pod,
+	claim *corev1.PersistentVolumeClaimVolumeSource,
+) (*corev1.PersistentVolumeClaim, *corev1.PersistentVolume, error) {
+	log := logf.FromContext(ctx).WithValues("pod", types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}, "pvc", claim.ClaimName)
+
+	pvc := &corev1.PersistentVolumeClaim{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: pod.Namespace, Name: claim.ClaimName}, pvc)
+	if err != nil {
+		log.Error(err, "Failed to get PVC for Pod")
+		return nil, nil, fmt.Errorf("Failed to get PVC for Pod: %w", err)
+	}
+
+	if pvc.Status.Phase != corev1.ClaimBound || pvc.Spec.VolumeName == "" {
+		log.V(debugLevel).Info("PVC is not bound to a PV yet or has a empty volume name - ignoring",
+			"status", pvc.Status.Phase,
+			"volumeName", pvc.Spec.VolumeName)
+		return nil, nil, errPVCIsNotBoundToAPV
+	}
+
+	pv := &corev1.PersistentVolume{}
+	err = r.Get(ctx, types.NamespacedName{Name: pvc.Spec.VolumeName}, pv)
+	if err != nil {
+		log.Error(err, "Failed to get PV bound to PVC", "volumeName", pvc.Spec.VolumeName)
+		return nil, nil, fmt.Errorf("Failed to get PV bound to PVC: %w", err)
+	}
+
+	if pv.Spec.ClaimRef == nil || pv.Spec.ClaimRef.Name != pvc.Name {
```

Comment:
Our recommendation is to ensure a PVC and a PV pre-bound to each other to ensure random PVCs and PVs won't get bound to each other unexpectedly. You can have a pre-bound, or otherwise PVCs and PVs can be bound to each other by Kubernetes by populating this `ClaimRef` field. Either way, this is the field on PV side that specifies which PVC its bound to (if any). Again if this field is nil or does refer to a different PVC than we expected, it's probably a bug in Kubernetes or any plugin modifies these objects.